tools/networkmonitor/networkmonitor_utils.nim

```diff
@@ -0,0 +1,47 @@
+import
+  std/[json,httpclient]
+import
+  chronicles,
+  chronicles/topics_registry,
+  chronos,
+  stew/results
+
+type
+  NodeLocation = object
+    country*: string
+    city*: string
+    lat*: string
+    long*: string
+    isp*: string
+
+proc flatten*[T](a: seq[seq[T]]): seq[T] =
+  result = @[]
+  for subseq in a:
+    result &= subseq
+
+# using an external api retrieves some data associated with the ip
+# TODO: use a cache
+proc ipToLocation*(ip: string,
+                   client: Httpclient):
+                   Future[Result[NodeLocation, string]] {.async.} =
+  # naive mechanism to avoid hitting the rate limit
+  # IP-API endpoints are now limited to 45 HTTP requests per minute
+  await sleepAsync(1400)
```

Comment:
If I understand correctly this will always be called for each discovered peer? I'm wondering what happens when the network scales to 1000s of nodes (here's hoping)? Would it make sense (perhaps only in a future PR) to split the monitoring into three different reporting procedures, spawned independently, that reports metrics on (1) what can be determined from discovered ENRs directly (no wait) (2) ipToLocation (with a period that respects the API limits) (3) connect to those nodes and report supported protocols? In other words the discovery loop (faster) is somewhat decoupled from the slower connection and location lookup procs?

Reply:
yup, definitely. i had something similar in mind :)